clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -164,25 +164,29 @@ public void abort(RuntimeException exception) {
      * @param exception The exception that occurred (or null if the request was successful)
      */
     public void done(long baseOffset, long logAppendTime, RuntimeException exception) {
-        final FinalState finalState;
-        if (exception == null) {
+        final FinalState tryFinalState = (exception == null) ? FinalState.SUCCEEDED : FinalState.FAILED;
+
+        if (tryFinalState == FinalState.SUCCEEDED) {
             log.trace("Successfully produced messages to {} with base offset {}.", topicPartition, baseOffset);
-            finalState = FinalState.SUCCEEDED;
         } else {
             log.trace("Failed to produce messages to {}.", topicPartition, exception);
-            finalState = FinalState.FAILED;
         }
 
-        if (!this.finalState.compareAndSet(null, finalState)) {
-            if (this.finalState.get() == FinalState.ABORTED) {
-                log.debug("ProduceResponse returned for {} after batch had already been aborted.", topicPartition);
-                return;
+        if (!this.finalState.compareAndSet(null, tryFinalState)) {
```

Comment:
The logic here probably needs more comments. We may have the following three cases that the state of a batch has been updated before the ProduceResponse returns:
1. A transaction abortion happens. The state of the batches would have been updated to `ABORTED`.
2. The producer is closed forcefully. The state of the batches would have been updated to `ABORTED`.
3. The batch is expired when it is in-flight. The state of the batch would have been updated to `EXPIRED`.

In the other cases, we should throw IllegalStateException.

Reply:
Please review the updated method documentation.